db/table_cache.cc

```diff
@@ -277,8 +277,17 @@ Status TableCache::Get(const ReadOptions& options,
     // sequence key increases. However, to support caching snapshot
     // reads, we append the sequence number (incremented by 1 to
     // distinguish from 0) only in this case.
-    uint64_t seq_no =
-        options.snapshot == nullptr ? 0 : 1 + GetInternalKeySeqno(k);
+    // If the snapshot is larger than the largest seqno in the file,
+    // all data should be exposed to the snapshot, so we treat it
+    // the same as there is no snapshot. The exception is that if
+    // a seqno checking callback is registered, some internal keys
+    // may still be filtered out.
+    uint64_t seq_no = 0;
+    if (options.snapshot != nullptr &&
+        (get_context->has_callback() ||
+         options.snapshot->GetSequenceNumber() <= fd.largest_seqno)) {
+      seq_no = 1 + GetInternalKeySeqno(k);
```

Comment:
You might consider using the snapshot seq number for clarity. To be on the safe side we could assert that it is equal to GetInternalKeySeqno as well but their equality is currently an implementation detail.

Reply:
I feel a little bit nervous about touching the code that is not related to the change. You may be right that it is easier to understand if it is from snapshot, but I don't want to take the risk of changing it. How do you feel keeping it as it is and add a comment? If you feel strongly about it, improve it in the next TechDebt week?